compiler/rustc_data_structures/src/sharded.rs

```diff
@@ -31,31 +35,53 @@ impl<T: Default> Default for Sharded<T> {
 impl<T> Sharded<T> {
     #[inline]
     pub fn new(mut value: impl FnMut() -> T) -> Self {
-        Sharded { shards: [(); SHARDS].map(|()| CacheAligned(Lock::new(value()))) }
+        Sharded {
+            #[cfg(parallel_compiler)]
+            mask: if is_dyn_thread_safe() { SHARDS - 1 } else { 0 },
+            shards: [(); SHARDS].map(|()| CacheAligned(Lock::new(value()))),
+        }
+    }
+
+    #[inline(always)]
+    fn mask(&self) -> usize {
+        #[cfg(parallel_compiler)]
+        {
+            self.mask
+        }
+        #[cfg(not(parallel_compiler))]
+        {
+            0
+        }
+    }
+
+    #[inline(always)]
+    fn count(&self) -> usize {
+        self.mask() + 1
     }
 
     /// The shard is selected by hashing `val` with `FxHasher`.
     #[inline]
     pub fn get_shard_by_value<K: Hash + ?Sized>(&self, val: &K) -> &Lock<T> {
-        if SHARDS == 1 { &self.shards[0].0 } else { self.get_shard_by_hash(make_hash(val)) }
+        self.get_shard_by_hash(if SHARDS == 1 { 0 } else { make_hash(val) })
```

Comment:
Should we return `&self.shards.get_unchecked(0).0` directly if `SHARDS == 1`?

Reply:
It gets masked by 0 later, so it doesn't matter.

Reply:
Could you make this a branch over `cfg(debug_assertions)`? I'd rather avoid having some logic over the cfg and some other over the value of `SHARDS`, even if they are equivalent.

Reply:
Assuming you mean `cfg(parallel_compiler)` here, it's still useful to disable just sharding by setting `SHARDS` to 1 to evaluate how effective sharding is.